src/Symfony/Component/Security/Core/Authentication/Provider/LdapQueryAuthenticationProvider.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Authentication\Provider;
+
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Core\Exception\UsernameNotFoundException;
+use Symfony\Component\Security\Core\User\UserCheckerInterface;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Ldap\LdapInterface;
+use Symfony\Component\Ldap\Exception\ConnectionException;
+
+/**
+ * LdapQueryAuthenticationProvider authenticates a user against an LDAP server.
+ *
+ * The only way to check user credentials is to first find the DN via query and then bind
+ *
+ * @author Charles Sarrazin <charles@sarraz.in>
```

Comment:
One thing, though. It seems that you have duplicate logic in the `else` part of your `if` statement:

```php
$username = $this->ldap->escape($username, '', LdapInterface::ESCAPE_DN);
```